src/psyclone/psyir/nodes/psy_data_node.py

```diff
@@ -0,0 +1,553 @@
+                       arguments)
+        parent.add(call)
+
+    # -------------------------------------------------------------------------
+    def gen_code(self, parent, options=None):
+        # pylint: disable=arguments-differ, too-many-branches
+        '''Creates the PSyData code before and after the children
+        of this node.
+
+        :param parent: the parent of this node in the f2pygen AST.
+        :type parent: :py:class:`psyclone.f2pygen.BaseGen`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
+        :param options["pre-var-list"]: a list of variables to be extracted \
+            before the first child.
+        :type options["pre-var-list"]: list of str
+        :param options["post-var-list"]: a list of variables to be extracted \
+            after the last child.
+        :type options["post-var-list"]: list of str
+        :param str options["pre-var-postfix"]: an optional postfix that will \
+            be added to each variable name in the pre-var-list.
+        :param str options["post-var-postfix"]: an optional postfix that will \
+            be added to each variable name in the post-var-list.
+
+        '''
+        # TODO: #415 Support different classes of PSyData calls.
+        module_name = self._module_name
+        if module_name is None:
+            # The user has not supplied a module (location) name so
+            # return the psy-layer module name as this will be unique
+            # for each PSyclone algorithm file.
+            module_name = self.root.invoke.invokes.psy.name
+
+        region_name = self._region_name
+        if region_name is None:
+            # The user has not supplied a region name (to identify
+            # this particular invoke region). Use the invoke name as a
+            # starting point.
+            region_name = self.root.invoke.name
+            from psyclone.psyGen import Kern
+            kerns = self.walk(Kern)
+            if len(kerns) == 1:
+                # This PSyData region only has one kernel within it,
+                # so append the kernel name.
+                region_name += ":{0}".format(kerns[0].name)
+            # Add a region index to ensure uniqueness when there are
+            # multiple regions in an invoke.
+            psy_data_nodes = self.root.walk(PSyDataNode)
+            idx = psy_data_nodes.index(self)
+            region_name += ":r{0}".format(idx)
+
+        if not options:
+            options = {}
+
+        pre_variable_list = options.get("pre-var-list", [])
+        post_variable_list = options.get("post-var-list", [])
+        pre_suffix = options.get("pre-var-postfix", "")
+        post_suffix = options.get("post-var-postfix", "")
+
+        # Note that adding a use statement makes sure it is only
+        # added once, so we don't need to test this here!
+        use = UseGen(parent, self.fortran_module, only=True,
```

Comment:
Hmm - I don't think it is required at this stage. We already do use the namespace manager for the name of the variable. All other variables cannot be used more than once, since the original code is already correct. E.g. the original code might use 'a_field' which is a dl_esm_info field. In the driver, the same name will be used, but as a normal 2d Fortran array. But the name is still unique and certain not to contract anything else.

We do need a namespace manager if we start supporting properties: a grid property like a_fld.%grid%nx (or so, can't remember the details) will be converted into a local variable called 'dx' ... and 'dx' might indeed be used in the program. This is already specified in #638 (todo item 3).

Reply:
We'll need to have registered e.g. "PSyDataType" with the namespace manager (or symboltable) at the start of the code-generation process. However I think #703 covers that.